RHDEVDOCS 5798 add performance information as requested by a customer (#69297)
openshift/openshift-docs: modules/op-improving-performance.adoc

// This module is included in the following assemblies:
// * resource/managing-pipelines-performance.adoc

:_content-type: PROCEDURE
[id="improving-performance_{context}"]
= Improving {pipelines-shortname} performance

If you experience slowness or recurrent failures of pipeline runs, you can take any of the following steps to improve the performance of {pipelines-shortname}.

> * Increase the number of nodes in the {OCP} cluster on which {pipelines-shortname} runs.

Reviewer: we need to word this differently so as to NOT draw a direct relation between increasing the nodes and performance improvement - performance will only improve when node resources (CPU, mem) are exhausted and only then the users should consider adding more nodes - this will be done as part of node resource monitoring separately and not much that tekton can help with

publish recommendation for number of nodes
- when running a large number of parallel workloads, resource consumption of nodes can become a bottleneck, hence it should be taken into consideration while observing openshift pipelines’ performance
- if nodes are running out of resources, increasing the number of nodes has shown to have significant impact in the failure rate of pipeline runs.

+
[NOTE]
====

> Increasing the CPU and RAM of individual nodes does not affect performance significantly.

Reviewer: i'm not sure if we can make this statement - there are times when node resources can become a performance bottleneck and can increase the workload failure rate - at that point increasing these resources will directly improve the performance

====

> * To enable the high-availability mode, in the `TektonConfig` custom resource (CR), set the `pipeline.performance.disable-ha` spec to `false`.

Reviewer: with HA enabled, we saw the following 2 metrics improve:
- time taken to complete pipelineruns was reduced by a significant amount
- taskrun creation to pod creation latency reduced by a significant amount

// This module is included in the following assemblies:
// * resource/managing-pipelines-performance.adoc

:_content-type: PROCEDURE
[id="improving-performance_{context}"]
= Improving {pipelines-shortname} performance

If you experience slowness or recurrent failures of pipeline runs, you can take any of the following steps to improve the performance of {pipelines-shortname}.

* Monitor the resource usage of the nodes in the {OCP} cluster on which {pipelines-shortname} runs. If the resource usage is high, increase the number of nodes.
+

> * Enable high-availability mode. This mode affects the controller that creates and starts pods for task runs and pipeline runs. In Red Hat testing, high-availability mode significantly reduced pipeline execution times as well as the delay from creating a `TaskRun` custom resource (CR) to the start of the pod executing the task run. To enable high-availability mode, make the following changes in the `TektonConfig` CR:

Reviewer: ```suggestion
* Enable high-availability mode. This mode affects the controller that creates and starts pods for task runs and pipeline runs. In Red Hat testing, high-availability mode significantly reduced pipeline execution times as well as the delay from creating a `TaskRun` resource to the start of the pod executing the task run. To enable high-availability mode, make the following changes in the `TektonConfig` CR:
```

> * Enable high-availability mode. This mode affects the controller that creates and starts pods for task runs and pipeline runs. In Red Hat testing, high-availability mode significantly reduced pipeline execution times as well as the delay from creating a `TaskRun` resource CR to the start of the pod executing the task run. To enable high-availability mode, make the following changes in the `TektonConfig` custom resource (CR):

Reviewer 2: I'm not very familiar with Pipelines, but this "procedure" seems to assume a lot of knowledge. While I am okay with approving this for now, I would like to see improvements in the future, showing an example CR and specs with callouts to where these fields are and how to adjust them. 

If you have time, this would be a great improvement to make today. Otherwise, I would prioritize improving this in the near term.

** Set the `pipeline.performance.disable-ha` spec to `false`.
** Set the `pipeline.performance.buckets` spec to a number between `5` and `10`.
** Set the `pipeline.performance.replicas` spec to a number higher than `2` and lower than or equal to the `pipeline.performance.buckets` setting.
+

> [NOTE]

Reviewer 2: The first half of this statement is a note, the second half is a warning. I would suggest breaking it up accordingly. Additionally, be more specific when referring to specs. See below:

```
[NOTE]
====
You can use different variables for `pipeline.performance.buckets` and `pipeline.performance.replicas` attributes to observe the effect on performance.
====

[WARNING]
====
Exhausting the node resources or the CPU and memory availability might occur. 
====

====

> You can try different numbers for buckets and replicas to observe the effect on performance. In general, higher numbers are eneficial if a large number of nodes is present in your {OCP} cluster.

Reviewer: ```suggestion
You can try different numbers for buckets and replicas to observe the effect on performance. In general, higher numbers are beneficial if a large number of nodes is present in your {OCP} cluster.
```

> You can try different numbers for buckets and replicas to observe the effect on performance. In general, higher numbers are eneficial if a large number of nodes is present in your {OCP} cluster.

Reviewer: ```suggestion
You can try different numbers for buckets and replicas to observe the effect on performance. In general, higher numbers are eneficial as long as you don't exhaust node resources OR as long as the CPU and memory resources on the node do not become a bottleneck.
```
